course_UvA-DL/contrastive-learning/SimCLR.py

```diff
@@ -0,0 +1,865 @@
+
+# %%
+img_transforms = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5, ), (0.5, ))])
+
+train_img_data = STL10(root=DATASET_PATH, split='train', download=True, transform=img_transforms)
+test_img_data = STL10(root=DATASET_PATH, split='test', download=True, transform=img_transforms)
+
+print("Number of training examples:", len(train_img_data))
+print("Number of test examples:", len(test_img_data))
+
+# %% [markdown]
+# Next, we implement a small function to encode all images in our datasets.
+# The output representations are then used as inputs to the Logistic Regression model.
+
+
+# %%
+@torch.no_grad()
+def prepare_data_features(model, dataset):
+    # Prepare model
+    network = deepcopy(model.convnet)
+    network.fc = nn.Identity()  # Removing projection head g(.)
+    network.eval()
+    network.to(device)
+
+    # Encode all images
+    data_loader = data.DataLoader(
+        dataset, batch_size=64, num_workers=NUM_WORKERS, shuffle=False, drop_last=False
+    )
+    feats, labels = [], []
+    for batch_imgs, batch_labels in tqdm(data_loader):
+        batch_imgs = batch_imgs.to(device)
+        batch_feats = network(batch_imgs)
+        feats.append(batch_feats.detach().cpu())
+        labels.append(batch_labels)
+
+    feats = torch.cat(feats, dim=0)
+    labels = torch.cat(labels, dim=0)
+
+    # Sort images by labels
+    labels, idxs = labels.sort()
+    feats = feats[idxs]
+
+    return data.TensorDataset(feats, labels)
+
+
+# %% [markdown]
+# Let's apply the function to both training and test set below.
+
+# %%
+train_feats_simclr = prepare_data_features(simclr_model, train_img_data)
+test_feats_simclr = prepare_data_features(simclr_model, test_img_data)
+
+# %% [markdown]
+# Finally, we can write a training function as usual.
+# We evaluate the model on the test set every 10 epochs to allow early
+# stopping, but the low frequency of the validation ensures that we do not
+# overfit too much on the test set.
+
+
+# %%
+def train_logreg(batch_size, train_feats_data, test_feats_data, model_suffix, max_epochs=100, **kwargs):
+    trainer = pl.Trainer(
+        default_root_dir=os.path.join(CHECKPOINT_PATH, "LogisticRegression"),
+        gpus=1 if str(device) == "cuda:0" else 0,
+        max_epochs=max_epochs,
+        callbacks=[
+            ModelCheckpoint(save_weights_only=True, mode='max', monitor='val_acc'),
+            LearningRateMonitor("epoch")
+        ],
+        progress_bar_refresh_rate=0,
+        check_val_every_n_epoch=10
+    )
+    trainer.logger._default_hp_metric = None
+
+    # Data loaders
+    train_loader = data.DataLoader(
+        train_feats_data,
+        batch_size=batch_size,
+        shuffle=True,
+        drop_last=False,
+        pin_memory=True,
+        num_workers=0
+    )
+    test_loader = data.DataLoader(
+        test_feats_data,
+        batch_size=batch_size,
+        shuffle=False,
+        drop_last=False,
+        pin_memory=True,
+        num_workers=0
+    )
```

Comment:
shall we wrap it into dataModule so then do not need to repeat the datalodder definition always...

Reply:
also the reader is ensured that the training data are identical and be wondering if some argument changed...